File: py/sorting.py
def selection_sort (arr):
    n = len(arr)
    for i in range(n):        
        p = i
        for j in range(i+1, n):
            if arr[j] < arr[p]:
                p = j
                
        tmp = arr[p];
        arr[p] = arr[i]
        arr[i] = tmp
                
def pivot_half(arr, start, end):
    return start + int((end -start) / 2)

def pivot_start(arr, start, end):
    return start

def pivot_median(arr, start, end):
    s = sum(arr) / (end - start);
    t = start;
    dt = float('inf')
    for i in range(start, end):
        d = abs(arr[i] - s)
        if d < dt:
            dt = d
            t = i
    return t 
          
def quick_sort(arr, pivotfunc):
    
    interval_stack = [[0,len(arr)]]
    while len(interval_stack) != 0:
        cur = interval_stack.pop(0)
    
        if(cur[0] >= cur[1]):
            continue
        
        p = arr[pivotfunc(arr, cur[0], cur[1])]
        write_idx = cur[0];
        for i in range(write_idx, cur[1]):
            if (arr[i] < p):
                arr[i], arr[write_idx] = arr[write_idx], arr[i]
                write_idx += 1
        
        mid = write_idx
        
        for i in range(write_idx, cur[1]):
            if(arr[i] == p):
                arr[i], arr[write_idx] = arr[write_idx], arr[i]
                write_idx += 1
        
        interval_stack.append([cur[0], mid])
        interval_stack.append([write_idx, cur[1]])
                

def native_sort(arr):
    arr.sort()
    

File: cpp/Sorting.cpp
#include "Sorting.h"

#include <iostream>
#include <chrono>
#include <deque>

size_t QuickSort(int arr[], size_t size)
{
  const auto start_time = std::chrono::high_resolution_clock::now();

  std::deque<std::pair<size_t, size_t>> interval_stack;
  interval_stack.push_back({ 0, size });

  while (!interval_stack.empty())
  {
    std::pair<size_t, size_t> cur = interval_stack.front();
    interval_stack.pop_front();

    if (cur.first >= cur.second)
      continue;

    const size_t pivot = arr[cur.first + ((cur.second - cur.first) / 2)];

    size_t write_idx = cur.first;

    for (size_t i = write_idx; i < cur.second; ++i)
      if (arr[i] < pivot)
        std::swap(arr[i], arr[write_idx++]);

    size_t mid = write_idx;

    for (size_t i = write_idx; i < cur.second; i++)
      if (arr[i] == pivot)
        std::swap(arr[i], arr[write_idx++]);

    interval_stack.push_back({ cur.first, mid });
    interval_stack.push_back({ write_idx, cur.second });
  }

  const auto end_time = std::chrono::high_resolution_clock::now();
  return std::chrono::duration_cast<std::chrono::nanoseconds>((end_time - start_time)).count();
}
